sql,logical: attribute ingestion queries to "external" things

yuzefovich opened the pull request:
**sqlstats: remove unused insights writer field**

This allows us to remove redundant `internal` boolean parameter in GetApplicationStats constructor.

**sql,logical: attribute ingestion queries to "external" things**

This commit introduces a way for the internally executed queries to be attributed to the user which is needed so that "external" observability would pick it up. In particular, we now can specify an InternalExecutorOverride option so that when instantiating the connExecutor we set the application name prefix to `$ public-internal` as well as we use the "server" metrics. As a result, such attributed internal queries will show up on the SQL Activity page as well as will be included into metrics like number of queries executed whenever this option is used. Note that these queries will keep contributing to "internal" things for less important stuff (like memory monitor metrics as well as txn guardrail metrics). No changes to the underlying internal executor machinery are applied so this _should_ be a pretty safe change.

The only hiccup I can think of right now is whether the hard-coded app name prefix that doesn't start with `$ internal` (or `$$` for "delegated" internal queries) can break some assumptions elsewhere.

This option is now utilized to attribute INSERT and DELETE queries of the ingestion pipeline of the LDR to the user. Newly added cluster setting `sql.internal_executor.attribute_to_user.enabled` can be used to disable this option.

Epic: CRDB-38996

Release note: None

yuzefovich committed:
sql,logical: attribute ingestion queries to "external" things

This commit introduces a way for the internally executed queries to be
attributed to the user which is needed so that "external" observability
would pick it up. In particular, we now can specify an
InternalExecutorOverride option so that when instantiating the
connExecutor we set the application name prefix to `$ public-internal`
as well as we use the "server" metrics. As a result, such attributed
internal queries will show up on the SQL Activity page as well as will
be included into metrics like number of queries executed whenever this
option is used. Note that these queries will keep contributing to
"internal" things for less important stuff (like memory monitor metrics
as well as txn guardrail metrics). No changes to the underlying internal
executor machinery are applied so this _should_ be a pretty safe change.

The only hiccup I can think of right now is whether the hard-coded app
name prefix that doesn't start with `$ internal` (or `$$` for
"delegated" internal queries) can break some assumptions elsewhere.

This option is now utilized to attribute INSERT and DELETE queries of
the ingestion pipeline of the LDR to the user. Newly added cluster
setting `sql.internal_executor.attribute_to_user.enabled` can be used to
disable this option.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125668)
<!-- Reviewable:end -->

yuzefovich commented:
@dhartunian @xinhaoz curious whether y'all can think of problems on the DB Console due to fiddling with the app name. The goal of this PR is to make it so that some internal queries would actually show up in the "user" space (like on SQL Activity page) without having to toggle the "include internal" cluster setting.

yuzefovich committed:
sqlstats: remove unused insights writer field

This allows us to remove redundant `internal` boolean parameter in
GetApplicationStats constructor.

Release note: None

rafiss reviewed (commented):
the overall structure lgtm, and thanks for adding that test! curious if the o11y team has thoughts.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125668#-:-O-Lzhg_6FKfNQum252y:b8gf7vj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @xinhaoz, and @yuzefovich)

___
*[`pkg/sql/internal.go` line 223 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125668#-O-LwubQCEnbsr44j5yy:-O-LwubQCEnbsr44j5yz:b3w93gb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1f3aa6f6943934539249e84a208379e678684d4/pkg/sql/internal.go#L223)):*
> ```Go
> 	syncCallback func([]*streamingCommandResult),
> 	errCallback func(error),
> 	attributeToUser bool,
> ```

i haven't looked extensively, but would it be possible to make this bool be a flag on the `InternalExecutor` struct, rather than a parameter here? (just to avoid adding too many params to the function)
___
*[`pkg/sql/sem/catconstants/constants.go` line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125668#-O-LxXer5mn-jbLv4vip:-O-LxXer5mn-jbLv4viq:bj0zcc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1f3aa6f6943934539249e84a208379e678684d4/pkg/sql/sem/catconstants/constants.go#L29)):*
> ```Go
> // identifies an internally-executed query that should be attributed to the
> // user.
> const AttributedToUserInternalAppNamePrefix = ReportableAppNamePrefix + "attributed-internal"
> ```

super nit: `attributed` might not be that intuitive to an end user who sees this name. perhaps a name like `$ public-internal` could convey the intended meaning?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125668)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, @rafiss, and @xinhaoz)

___
*[`pkg/sql/internal.go` line 223 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125668#-O-LwubQCEnbsr44j5yy:-O-MfUeZ-Q5d5X-ssCPA:b-vgk5kp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1f3aa6f6943934539249e84a208379e678684d4/pkg/sql/internal.go#L223)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i haven't looked extensively, but would it be possible to make this bool be a flag on the `InternalExecutor` struct, rather than a parameter here? (just to avoid adding too many params to the function)
</blockquote></details>

Making this flag be a part of `InternalExecutor` struct seems like it could be racy. I think that the same struct can be shared by multiple components running concurrently, so there could be races between multiple calls to `Exec*` methods overriding this flag.
___
*[`pkg/sql/sem/catconstants/constants.go` line 29 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/125668#-O-LxXer5mn-jbLv4vip:-O-MgCef421ovybvBw8c:b-86mr6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1f3aa6f6943934539249e84a208379e678684d4/pkg/sql/sem/catconstants/constants.go#L29)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

super nit: `attributed` might not be that intuitive to an end user who sees this name. perhaps a name like `$ public-internal` could convey the intended meaning?
</blockquote></details>

I like it, done.


<!-- Sent from Reviewable.io -->

xinhaoz commented on pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go:
```diff
@@ -97,7 +97,6 @@ type Container struct {
 	mon       *mon.BytesMonitor
 
 	knobs              *sqlstats.TestingKnobs
-	insights           insights.Writer
```
Oops, missed this. TY for deleting!

xinhaoz reviewed (commented):
LGTM! The only thing I can think of is if it's used to surface queries with outer txns they won't have an equivalent entry in the transaction stats page, if they try to look for that -https://github.com/cockroachdb/cockroach/issues/124935. Unsure if we want to try to explain that in the UI or not until we address that ticket.

Also @dhartunian perhaps we should extend the SQL api to have this flag for the cloud sql shell? Those queries seem to fall into this category and would be nice to surface as user queries.

yuzefovich reviewed (commented):
Ack on #124935 - it sounds like an acceptable limitation to me, at least for now.

Using this new flag for could sql shell queries also makes sense to me, but I'll leave it to you to do as a follow-up.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125668#-:-O-auHPT5gtdmkEI22MG:b-mwz87q)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @dt, and @rafiss)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
I added a cluster setting as an escape hatch for this in case we want to revert to the previous behavior. Will go ahead and merge this to help LDR effort.

TFTRs!

bors r+

yuzefovich commented:
Oh, I guess neither @rafiss nor @xinhaoz gave a GitHub approval to this.

yuzefovich commented:
Thanks!

bors r+
